src/modules/flight_mode_manager/tasks/AutoFollowTarget/FlightTaskAutoFollowTarget.cpp

```diff
@@ -43,330 +43,386 @@
-	if (!PX4_ISFINITE(_target_position_filter.getState()(0)) || !PX4_ISFINITE(_target_position_filter.getState()(1))
-	    || !PX4_ISFINITE(_target_position_filter.getState()(2))) {
-		_target_position_filter.reset(pos_ned_est);
-	}
+	float gimbal_height{0.0f};
 
-	// Low-pass filter on target position.
-	_target_position_filter.setParameters(_deltatime, POSITION_FILTER_ALPHA);
+	switch (altitude_mode) {
+		case kFollowAltitudeModeTerrain:
+			// Point the gimbal at the ground level in this tracking mode
+			gimbal_height = _dist_to_ground;
+			break;
 
-	if (_param_nav_ft_delc.get() == 0) {
-		_target_position_filter.update(pos_ned_est);
+		case kFollowAltitudeModeTrackTarget:
+			// Point the gimbal at the target's 3D coordinates
+			gimbal_height = -(_position(2) - target_pos_z);
+			break;
 
-	} else {
-		// Use a predicted target's position to compensate the filter delay to some extent.
-		const Vector3f target_predicted_position = predict_future_pos_ned_est(POSITION_FILTER_ALPHA, pos_ned_est, vel_ned_est,
-				acc_ned_est);
-		_target_position_filter.update(target_predicted_position);
-	}
+		case kFollowAltitudeModeConstant:
 
-	return _target_position_filter.getState();
+		//FALLTHROUGH
 
+		default:
+			gimbal_height = _home_position_z - _position(2); // Assume target is at home position's altitude
+			break;
+	}
+
+	return gimbal_height;
 }
 
+
 bool FlightTaskAutoFollowTarget::update()
 {
-	_follow_target_estimator_sub.update(&_follow_target_estimator);
 	follow_target_status_s follow_target_status{};
+	bool in_emergency_ascent{false};
+	float gimbal_pitch{NAN};
+	float raw_orbit_angle_setpoint{NAN};
+
+	// Get the latest target estimator message for target position and velocity
+	_follow_target_estimator_sub.update(&_follow_target_estimator);
 
 	if (_follow_target_estimator.timestamp > 0 && _follow_target_estimator.valid) {
-		const Vector3f pos_ned_est{_follow_target_estimator.pos_est};
-		const Vector3f vel_ned_est{_follow_target_estimator.vel_est};
-		const Vector3f acc_ned_est{_follow_target_estimator.acc_est};
-		const Vector3f target_position_filtered = calculate_target_position_filtered(pos_ned_est, vel_ned_est, acc_ned_est);
-		const Vector2f offset_vector_filtered = calculate_offset_vector_filtered(vel_ned_est);
-		const Vector3f drone_desired_position = calculate_drone_desired_position(target_position_filtered,
-							offset_vector_filtered);
+		updateTargetPositionVelocityFilter(_follow_target_estimator);
+		const Vector3f target_position_filtered = _target_position_velocity_filter.getState();
+		const Vector3f target_velocity_filtered = _target_position_velocity_filter.getRate();
+		const Vector2f drone_to_target_vector  = Vector2f(target_position_filtered.xy()) - Vector2f(_position.xy());
+
+		// Calculate heading to the target (for Yaw setpoint)
+		const float drone_to_target_heading = atan2f(drone_to_target_vector(1), drone_to_target_vector(0));
```

Comment:
Something to test, but you might get a more stable yaw setpoint if you use the heading from the position setpoint to the target estimate. The drone should never be far from the position setpoint anyway.

Reply:
Yes and no, I think the yaw control will be smoother but when the target orientation changes rapidly for example, I think it could lag quite a lot and degrade capability of tracking the target in the center. Something to test!

Reply:
Added to the notes in the PR!

Reply:
I added my test result & thoughts in the [comment ](https://github.com/PX4/PX4-Autopilot/pull/19260#issuecomment-1122821572)below. My conclusion is that keeping it as-is is better!

Reply:
Thanks for the testing, I agree with your conclusion